### PR TITLE
Taking a raw ref (`&raw (const|mut)`) of a deref of pointer (`*ptr`) is always safe

### PR DESCRIPTION
**Pull request by compiler-errors**
T-opsem decided in https://github.com/rust-lang/reference/pull/1387 that `*ptr` is only unsafe if the place is accessed. This means that taking a raw ref of a deref expr is always safe, since it doesn't constitute a read.

This also relaxes the `DEREF_NULLPTR` lint to stop warning in the case of raw ref of a deref'd nullptr, and updates its docs to reflect that change in the UB specification.

This does not change the behavior of `addr_of!((*ptr).field)`, since field projections still require the projection is in-bounds.

I'm on the fence whether this requires an FCP, since it's something that is guaranteed by the reference you could ostensibly call this a bugfix since we were counting truly safe operations as unsafe. Perhaps someone on opsem has a strong opinion? cc @rust-lang/opsem

### CODE REVIEW

**Comment by RalfJung** [comment]
Seems more like a lang question to me than an opsem question -- from an opsem side, the relevant question is whether there is UB, and the answer is "no". But from a lang perspective, it may seem odd to have `unsafe` rules that depend on the place-vs-value distinction, as many programmers do not properly understand this distinction. But maybe it's good that they will wonder " :thinking: why is this safe?" and then can be taught about said distinction?

Also note that due to https://github.com/rust-lang/rust/issues/117288, sometimes we do have UB just from constructing a place. That's a MIR building bug but we should probably fix it before accepting such code as safe. (Sorry I didn't remember this when we spoke earlier.)

**Comment by compiler-errors** [comment]
👍 yep, this could be blocked on #117288 then. I think we can probably T-lang-nominate this regardless for discussion.

**Comment by nnethercote** [comment]
The code changes seem fine, so r=me on them if/when lang/opsem are happy with this.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in lang triage today and agreed that this was both a good idea and one that's a natural consequence of our other recent decisions and so can go without FCP.

@bors r=nnethercode

**Comment by compiler-errors** [comment]
@bors r-

This is still unsound!! https://github.com/rust-lang/rust/issues/117288

**Comment by traviscross** [comment]
Good catch.  Thanks.

File: tests/ui/static/raw-ref-deref-without-unsafe.rs
Comment (workingjubilee) [inline tests/ui/static/raw-ref-deref-without-unsafe.rs:1]: This has come to pass, and so the remark should be altered appropriately:
```rust
// even if in a hypothetical future we deem all &raw (const|mut) *ptr exprs safe.
```

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 1]
```diff
@@ -503,17 +503,11 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 // THIR desugars UNSAFE_STATIC into *UNSAFE_STATIC_REF, where
                 // UNSAFE_STATIC_REF holds the addr of the UNSAFE_STATIC, so: take two steps
```
Comment (workingjubilee) [inline compiler/rustc_mir_build/src/check_unsafety.rs:504]: Trimmed?

Comment (compiler-errors) [reply compiler/rustc_mir_build/src/check_unsafety.rs:504]: Ah, I need to update this comment. We don't care about statics here anymore.

**Comment by workingjubilee** [comment]
> warning: the feature `raw_ref_op` has been stable since 1.82.0-nightly and no longer requires an attribute to enable

Truly?!

**Comment by RalfJung** [comment]
Yes, truly. :)

https://github.com/rust-lang/rust/pull/127679

**Comment by RalfJung** [comment]
So `let _ = *raw_ptr;` and `match *raw_ptr { _ => ... }` are not affected by this?

That does seem a bit inconsistent. It doesn't seem super useful to write code like that, but then it doesn't seem super useful to write `&raw const *ptr` either.

**Comment by compiler-errors** [comment]
@RalfJung: No, that would need to have another special case in the THIR.

**Comment by compiler-errors** [comment]
> That does seem a bit inconsistent. It doesn't seem super useful to write code like that, but then it doesn't seem super useful to write &raw const *ptr either.

🤷 I'd still like to fix it, regardless, I think, especially now that #117288 is closed. I'd also like if `let _ = *raw_ptr;` weren't unsafe, but that's far more difficult for unsafeck to implement.

**Comment by compiler-errors** [comment]
Because it was unsound? Could you provide some context.

**Comment by compiler-errors** [comment]
also this is 

@rustbot ready 

:D

**Comment by nnethercote** [comment]
@RalfJung, @scottmcm: any objections before I approve this?

**Comment by RalfJung** [comment]
Procedurally it has been [approved](https://github.com/rust-lang/rust/pull/129248#issuecomment-2302506438) by t-lang. I'm not very familiar with THIR so it's hard for me to say whether the implementation matches the intent. We don't have a dedicated test for the cases that should be rejected, but tests/ui/lint/lint-deref-nullptr.rs and tests/ui/static/raw-ref-deref-without-unsafe.rs to cover the most important cases I can think of.

**Comment by nnethercote** [comment]
I think this has waited long enough.

@bors r+

**Pull request merged**